cpp/calculate-demand.cpp

```diff
@@ -0,0 +1,269 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <math.h>
+#include <stdio.h>
+
+#include "gdal_priv.h"
+#include "cpl_conv.h"
+#include "cpl_string.h"
+
+typedef unsigned char BYTE;
+
+#define DEBUG 1
+
+const char* DEMAND_METADATA_KEY = "UNSATISFIED_DEMAND";
+const char* APP_METADATA_DOMAIN = "PLANWISE";
+
+// See http://stackoverflow.com/a/13636164/12791
+template <typename T> std::string numberToString (T number) {
+  std::ostringstream stringStream;
+  stringStream << number;
+  return stringStream.str();
+}
+
+bool fileExists (const std::string& name) {
+  std::ifstream fileStream(name.c_str());
+  return fileStream.good();
+}
+
+GDALDataset* openRaster(std::string filename) {
+  GDALDataset* poDataset = (GDALDataset*) GDALOpen(filename.c_str(), GA_ReadOnly);
+  if (poDataset == NULL) {
+    std::cerr << "Failed opening: " << filename << std::endl;
+    exit(1);
+  }
+  return poDataset;
+};
+
+GDALDataset* copyRaster(GDALDataset* sourceRaster, std::string filename, int blockXSize, int blockYSize) {
+  const char *pszFormat = "GTiff";
+  GDALDriver *poDriver;
+  poDriver = GetGDALDriverManager()->GetDriverByName(pszFormat);
+  if (poDriver == NULL) {
+    std::cerr << "Failed loading driver" << std::endl;
+    exit(1);
+  }
+
+  char **papszOptions = NULL;
+  papszOptions = CSLSetNameValue(papszOptions, "TILED", "YES");
+  papszOptions = CSLSetNameValue(papszOptions, "BLOCKXSIZE", numberToString(blockXSize).c_str());
+  papszOptions = CSLSetNameValue(papszOptions, "BLOCKYSIZE", numberToString(blockYSize).c_str());
+
+  GDALDataset* dataset = poDriver->CreateCopy(filename.c_str(), sourceRaster, FALSE, papszOptions, NULL, NULL);
+  dataset->FlushCache();
+  return dataset;
+}
+
+void closeRaster(GDALDataset* rasterDataSet) {
+  GDALClose(rasterDataSet);
+};
+
+long readUnsatisfiedDemand(std::string targetFilename) {
+  GDALDataset* targetDataset = openRaster(targetFilename);
+  const char* metadataValue = targetDataset->GetMetadataItem(DEMAND_METADATA_KEY, APP_METADATA_DOMAIN);
+  if (metadataValue == NULL) {
+    std::cerr << "No unsatisfied demand metadata found on " << APP_METADATA_DOMAIN << ":" << DEMAND_METADATA_KEY << std::endl;
+    exit(1);
+  }
+  closeRaster(targetDataset);
+  return atol(metadataValue);
+}
+
+long calculateUnsatisfiedDemand(std::string targetFilename, std::string demoFilename, std::vector<std::string> facilities, std::vector<float> capacities) {
+  GDALDataset* demoDataset = openRaster(demoFilename);
+  GDALRasterBand* demoBand = demoDataset->GetRasterBand(1);
+  CPLAssert(demoBand->GetRasterDataType() == GDT_Float32);
```

Comment:
It depends on how catastrophic is that the assertion is not valid. For user input validation (as in program parameters in this case) aborting with `exit(1)` seems the right choice here.